sklearn/metrics/pairwise.py

```diff
@@ -917,20 +932,27 @@ def gower_distances(X, Y=None, categorical_features=None, scale=True):
         if X.shape[0] != Y.shape[0] or X is not Y:
             j_start = 0
 
-        # Calculates the similarities for categorical columns
-        cat_dists = (X[i, cat_mask] != Y[j_start:, cat_mask]) | \
-                    (_object_dtype_isnan(X[i, cat_mask]) |
-                     _object_dtype_isnan(Y[j_start:, cat_mask]))
+        # Makes the comparisson for np.nan for arrays with dtype=np.object,
+        # this is necessary as some deployments returns True for
+        # np.nan == np.nan
+        cat_nan_cols = (_object_dtype_isnan(X[i, cat_mask]) |
+                        _object_dtype_isnan(Y[j_start:, cat_mask]))
 
+        # Calculates the similarities for categorical columns
+        cat_dists = ((X[i, cat_mask] != Y[j_start:, cat_mask]) | cat_nan_cols)
         # Calculates the Manhattan distances for numerical columns
-        num_dists = abs(X[i, num_mask].astype(np.float32) -
-                        Y[j_start:, num_mask].astype(np.float32)) / ranges
+        num_dists = abs(X[i, num_mask] -
+                        Y[j_start:, num_mask]) / ranges
+
         # Calculates the number of non missing columns
-        n_missing = X.shape[1] - (np.isnan(cat_dists).sum(axis=1) +
-                                  np.isnan(num_dists).sum(axis=1))
+        n_missing = X.shape[1] - (cat_nan_cols.sum(axis=1) +
+                                  _object_dtype_isnan(num_dists).sum(axis=1))
         # Gets the final results
-        results = np.sum(cat_dists, axis=1) + np.sum(num_dists, axis=1)
-        results /= n_missing
+        sum = np.sum(cat_dists, axis=1) + np.sum(num_dists, axis=1)
+
+        results = np.divide(sum, n_missing,
+                            out=np.full(sum.shape, np.nan, dtype=np.object),
+                            where=n_missing != np.zeros(sum.shape))
```

Comment:
Is this just here to broadcast `n_missing != 0`??

Reply:
Done